Implements the second half of dragway::RoadGeometry::DoToRoadPosition().

liangfok opened the pull request:
This PR builds upon #5313 by handling cases where the geographic position is not on top of the dragway's driveable region.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5320)
<!-- Reviewable:end -->

liangfok commented:
+@jadecastro for feature review please.
+@sammy-tri for platform review please.
+@jwnimmer-tri for car review please.

Thanks!

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-:-Ke8fOCNn9y_6oO2jH42:b-200czk)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@soonho-tri for platform review please.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-:-Ke8uBIG1jGqDJyVyO56:b-vt9vbo)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot linux-clang-experimental-debug please.

jwnimmer-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-Ke9LqCMUAEgD8zvleUp:-Ke9LqCMUAEgD8zvleUq:bptvn46) (no related file):*
So the conceptual change is fine.  I do worry though about the bloat of the code and unit test.  Shouldn't it just be about two lines long?
```c++
result.x = clamp(x, min_x, max_x);
result.y = clamp(y, min_y, max_y);
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-Ke9LqCMUAEgD8zvleUp:-Ke9N_q_pgZz0pjtbV5S:b-cnceq4) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

So the conceptual change is fine.  I do worry though about the bloat of the code and unit test.  Shouldn't it just be about two lines long?
```c++
result.x = clamp(x, min_x, max_x);
result.y = clamp(y, min_y, max_y);
```
</blockquote></details>

I think that might work. Let me test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput/dragway/road_geometry.cc, line 228 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-Ke9OSRilmrIjxsOfy4Q:-Ke9OSRilmrIjxsOfy4R:b-jh72wi) ([raw file](https://github.com/robotlocomotion/drake/blob/da9fefb821e41b038e5fcbbca1b4fee65090329f/drake/automotive/maliput/dragway/road_geometry.cc#L228)):*
> ```cpp
> 
>   if (!region_found) {
>     throw std::runtime_error("dragway::RoadGeometry::GetClosestRoadPosition: "
> ```

FYI, this branch is not covered by unit tests.z

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 396 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-Ke9PksALeyHYNKEWvVD:-Ke9PksALeyHYNKEWvVE:b-7dizuf) ([raw file](https://github.com/robotlocomotion/drake/blob/da9fefb821e41b038e5fcbbca1b4fee65090329f/drake/automotive/maliput/dragway/test/dragway_test.cc#L396)):*
> ```cpp
>             std::pow(y + y_min, 2));
>         EXPECT_DOUBLE_EQ(distance, expected_distance);
>         const api::Lane* expected_lane =
> ```

FYI, you have `const api::Lane*` without using `dynamic_cast` here and in the test for region 6, while you have `const Lane*` with `dynamic_cast` for other tests.

META: probably better to have a parameterized test to avoid repetition?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/automotive/maliput/dragway/road_geometry.h, line 94 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-Ke9PmTGX-sPlZgmYhxv:-Ke9PmTGX-sPlZgmYhxw:briahdh) ([raw file](https://github.com/robotlocomotion/drake/blob/da9fefb821e41b038e5fcbbca1b4fee65090329f/drake/automotive/maliput/dragway/road_geometry.h#L94)):*
> ```objc
>   api::RoadPosition GetClosestRoadPosition(
>     const api::GeoPosition& geo_pos,
>     const api::RoadPosition* hint,
> ```

BTW, the function doesn't appear to override the API, so you can probably get rid of `hint` since you never use it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320)*
<!-- Sent from Reviewable.io -->

liangfok commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-Ke9LqCMUAEgD8zvleUp:-Ke9UuaucgbmbsYNtcLe:b-za4lxz) (no related file):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I think that might work. Let me test.
</blockquote></details>

My initial testing indicates this is feasible. I shall thus mark this PR "do not review" until I revise it using the clamping method. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks!

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-Ke9LqCMUAEgD8zvleUp:-Ke9clcuz8s7TaHPAv7f:b-896fix) (no related file):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

My initial testing indicates this is feasible. I shall thus mark this PR "do not review" until I revise it using the clamping method. Thanks!
</blockquote></details>

Done.

---

*[drake/automotive/maliput/dragway/road_geometry.cc, line 228 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-Ke9OSRilmrIjxsOfy4Q:-Ke9cRxMqk9y93idzS05:b-bh2luu) ([raw file](https://github.com/robotlocomotion/drake/blob/da9fefb821e41b038e5fcbbca1b4fee65090329f/drake/automotive/maliput/dragway/road_geometry.cc#L228)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, this branch is not covered by unit tests.z
</blockquote></details>

Done. I removed this branch.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 396 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-Ke9PksALeyHYNKEWvVD:-Ke9cXk0d2r5NeYx1smn:b-c2yxrq) ([raw file](https://github.com/robotlocomotion/drake/blob/da9fefb821e41b038e5fcbbca1b4fee65090329f/drake/automotive/maliput/dragway/test/dragway_test.cc#L396)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, you have `const api::Lane*` without using `dynamic_cast` here and in the test for region 6, while you have `const Lane*` with `dynamic_cast` for other tests.

META: probably better to have a parameterized test to avoid repetition?
</blockquote></details>

Done. I simplified the test to simply search a region that is a superset of the driveable region along both the x and y axes.

---

*[drake/automotive/maliput/dragway/road_geometry.h, line 94 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-Ke9PmTGX-sPlZgmYhxv:-Ke9ci-iq_9sIomv0VAl:b21ttob) ([raw file](https://github.com/robotlocomotion/drake/blob/da9fefb821e41b038e5fcbbca1b4fee65090329f/drake/automotive/maliput/dragway/road_geometry.h#L94)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, the function doesn't appear to override the API, so you can probably get rid of `hint` since you never use it?
</blockquote></details>

Done. I removed this method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-:-Ke9cmGJP3iluHujfmJK:bblnx7i)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Reviewed 1 of 3 files at r2.
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/dragway/road_geometry.cc, line 175 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-Ke9dEZ6c2qPEzXVMVSm:-Ke9dEZ6c2qPEzXVMVSn:b-k4tb8c) ([raw file](https://github.com/robotlocomotion/drake/blob/337d49ead5b621a9a6f029b7d167fa6061a94a6a/drake/automotive/maliput/dragway/road_geometry.cc#L175)):*
> ```cpp
>                   |         :         |
>                             :
>                             V
> ```

BTW, does it make sense to keep this illustration anymore?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/dragway/road_geometry.cc, line 175 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-Ke9dEZ6c2qPEzXVMVSm:-Ke9f2jYVQ3KlLYL2fC7:bheehkz) ([raw file](https://github.com/robotlocomotion/drake/blob/337d49ead5b621a9a6f029b7d167fa6061a94a6a/drake/automotive/maliput/dragway/road_geometry.cc#L175)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, does it make sense to keep this illustration anymore?
</blockquote></details>

I think it will help future readers visualize the semantics of `x_min`, `x_max`, `y_min`, and `y_max`. No?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> automotive sanity check.

I do think the same "should we have a set of 10 hard-coded test points" versus "lots of loops and reimplementing the code" discussion from the prior PR still stands, but will let Jon iterate on that.

---

Reviewed 1 of 3 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-:-Ke9fDEBUuOTejWpZpRB:bh6yx7l)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-:-KeAEqT-9DcGbEddRfnD:bnfp4nl)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-:-KeAFqOeoyLUoxzTZNyf:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Again, the test is not complete, so more points will not buy you anything in terms of code-path coverage. I think you can get away getting rid of the `/4`in your loop incrementer so that you only cover four `z=0` cases.  And, you only need one additional nonzero `z` value at any of these cases to check that `distance` does not misbehave for that instance. 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 422 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-KeAH9RzabNisZ2P-PSl:-KeAH9S-mU1OerEQbN7P:bnq04ue) ([raw file](https://github.com/robotlocomotion/drake/blob/086fdc9e3befff706bb39374bace7285a46acbe9/drake/automotive/maliput/dragway/test/dragway_test.cc#L422)):*
> ```cpp
>             std::pow(y - expected_nearest_position.y, 2) +
>             std::pow(z - expected_nearest_position.z, 2));
>         EXPECT_DOUBLE_EQ(distance, expected_distance);
> ```

BTW, I think all you need to do is to check here that `distance > 0`. The `expected_distance` is really rehashing the formula inside your function, so it seems brittle to your particular implementation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-:-KeACrZ5hmS6UqaoTzVn:bmwuc1q)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 378 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-KeAK9Sx7ux6_8XNizVT:-KeAK9Sx7ux6_8XNizVU:bbhuix9) ([raw file](https://github.com/robotlocomotion/drake/blob/086fdc9e3befff706bb39374bace7285a46acbe9/drake/automotive/maliput/dragway/test/dragway_test.cc#L378)):*
> ```cpp
>   // Spot checks a region that is a superset of the road's driveable bounds with
>   // a focus on edge cases.
>   const double test_x_min = x_min - 10;
> ```

Rather than a large value like `10`, I would think you'd want a small value like `1e-6` to check if it works when outside the dragway by some small `epsilon`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

---

Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 378 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-KeAK9Sx7ux6_8XNizVT:-KeATN_EQRRH5f3opItL:b-iwvk2l) ([raw file](https://github.com/robotlocomotion/drake/blob/086fdc9e3befff706bb39374bace7285a46acbe9/drake/automotive/maliput/dragway/test/dragway_test.cc#L378)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Rather than a large value like `10`, I would think you'd want a small value like `1e-6` to check if it works when outside the dragway by some small `epsilon`?
</blockquote></details>

Done. I now test points that are `1e-10` beyond the dragway's driveable region's x and y directions.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 422 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-KeAH9RzabNisZ2P-PSl:-KeARfzmkFXGmlnaUBI6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/086fdc9e3befff706bb39374bace7285a46acbe9/drake/automotive/maliput/dragway/test/dragway_test.cc#L422)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, I think all you need to do is to check here that `distance > 0`. The `expected_distance` is really rehashing the formula inside your function, so it seems brittle to your particular implementation.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-:-KeARMTtU_j7rUgWUcl9:bblnx7v)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Great! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  

---

Reviewed 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-:-KeAUhjSlPiE11utkcnK:b-3xptcd)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Thanks! Now waiting on CI.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5320#-:-KeAWA0FxfhJR7qgeZ5G:b-5y5ebz)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot linux-gcc-experimental-ros please.
